instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb

```diff
@@ -0,0 +1,155 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'lru_redux'
+require_relative '../constants'
+
+module OpenTelemetry
+  module Instrumentation
+    module PG
+      module Patches
+        # Module to prepend to PG::Connection for instrumentation
+        module Connection
+          PG::Constants::EXEC_ISH_METHODS.each do |method|
+            define_method method do |*args|
+              span_name, attrs = span_attrs(:query, *args)
+              tracer.in_span(span_name, attributes: attrs, kind: :client) do
+                super(*args)
+              end
+            end
+          end
+
+          PG::Constants::PREPARE_ISH_METHODS.each do |method|
+            define_method method do |*args|
+              span_name, attrs = span_attrs(:prepare, *args)
+              tracer.in_span(span_name, attributes: attrs, kind: :client) do
+                super(*args)
+              end
+            end
+          end
+
+          PG::Constants::EXEC_PREPARED_ISH_METHODS.each do |method|
+            define_method method do |*args|
+              span_name, attrs = span_attrs(:execute, *args)
+              tracer.in_span(span_name, attributes: attrs, kind: :client) do
+                super(*args)
+              end
+            end
+          end
+
+          private
+
+          def tracer
+            PG::Instrumentation.instance.tracer
+          end
+
+          def config
+            PG::Instrumentation.instance.config
+          end
+
+          def lru_cache
+            # When SQL is being sanitized, we know that this cache will
+            # never be more than 50 entries * 2000 characters (so, presumably
+            # 100k bytes - or 97k). When not sanitizing SQL, then this cache
+            # could grow much larger - but the small cache size should otherwise
+            # help contain memory growth. The intended use here is to cache
+            # prepared SQL statements, so that we can attach a reasonable
+            # `db.sql.statement` value to spans when those prepared statements
+            # are executed later on.
+            @lru_cache ||= ::LruRedux::ThreadSafeCache.new(50)
+          end
```

Comment:
I'm not sure if 50 is the right number here - I mostly just made it up, did the math, and thought "gee that seems fine."